Lib/unittest/mock.py

```diff
@@ -2346,25 +2347,12 @@ def __init__(self, spec, spec_set=False, parent=None,
 
 file_spec = None
 
-def _iterate_read_data(read_data):
-    # Helper for mock_open:
-    # Retrieve lines from read_data via a generator so that separate calls to
-    # readline, read, and readlines are properly interleaved
-    sep = b'\n' if isinstance(read_data, bytes) else '\n'
-    data_as_list = [l + sep for l in read_data.split(sep)]
-
-    if data_as_list[-1] == sep:
-        # If the last line ended in a newline, the list comprehension will have an
-        # extra entry that's just a newline.  Remove this.
-        data_as_list = data_as_list[:-1]
-    else:
-        # If there wasn't an extra newline by itself, then the file being
-        # emulated doesn't have a newline to end the last line  remove the
-        # newline that our naive format() added
-        data_as_list[-1] = data_as_list[-1][:-1]
 
-    for line in data_as_list:
-        yield line
+def _to_stream(read_data):
+    if isinstance(read_data, bytes):
```

Comment:
Code LGTM. I am slightly concerned if there might be any problems with backport where in Python 2 bytes is like an alias for str. So, even with BytesIO it would still be string as shown below and work as expected. Maybe I am just overthinking this. 

It would be helpful if @cjw296 and @voidspace can review this since this is a good bug to be fixed. Thanks @remilapeyre .

```
$ python2
Python 2.7.6 (default, Sep  9 2014, 15:04:36)
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.39)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> isinstance('a', bytes)
True
>>> isinstance(b'a', str)
True
>>> import io
>>> io.BytesIO('a').readline()
'a'
>>> type(io.BytesIO('a').readline())
<type 'str'>
```

Reply:
Don't worry, we can make this work in the backport :-)

Reply:
@tirkarthi wouldn't it still be correct? The test would be useless be the mock would still work right?

Reply:
yes, in Python 2 str and bytes are the same. I will just keep on the CI once the backport lands in mock repo.